src/value.rs

```diff
@@ -0,0 +1,181 @@
+        py: Python<'py>,
+    ) -> Result<PyDeserializedValue<'py>, DeserializationError>;
+}
+
+impl<'frame, 'metadata, 'py, T> PyDeserializeValue<'frame, 'metadata, 'py> for Vec<T>
+where
+    T: PyDeserializeValue<'frame, 'metadata, 'py>,
+{
+    fn deserialize_py(
+        typ: &'metadata ColumnType<'metadata>,
+        v: Option<FrameSlice<'frame>>,
+        py: Python<'py>,
+    ) -> Result<PyDeserializedValue<'py>, DeserializationError> {
+        let elem_typ = match typ {
+            ColumnType::Collection {
+                frozen: false,
+                typ: CollectionType::List(elem_typ),
+            }
+            | ColumnType::Collection {
+                frozen: false,
+                typ: CollectionType::Set(elem_typ),
+            } => elem_typ,
+            _ => {
+                unreachable!("Should not happen")
+            }
+        };
+
+        let mut v = if let Some(v) = v {
+            v
+        } else {
+            return Ok(PyDeserializedValue::new(PyList::empty(py).into_any()));
+        };
+
+        let count = types::read_int_length(v.as_slice_mut()).map_err(DeserializationError::new)?;
+
+        let raw_iter = FixedLengthBytesSequenceIterator::new(count, v);
+
+        let list = PyList::empty(py);
+        for raw_opt in raw_iter {
+            let raw = raw_opt.map_err(DeserializationError::new)?;
+            let item = T::deserialize_py(elem_typ, raw, py)?;
+            list.append(item).map_err(DeserializationError::new)?;
+        }
+
+        Ok(PyDeserializedValue::new(list.into_any()))
+    }
+}
+
+impl<'frame, 'metadata, 'py, T> PyDeserializeValue<'frame, 'metadata, 'py> for Option<T>
+where
+    T: PyDeserializeValue<'frame, 'metadata, 'py>,
+{
+    fn deserialize_py(
+        typ: &'metadata ColumnType<'metadata>,
+        v: Option<FrameSlice<'frame>>,
+        py: Python<'py>,
+    ) -> Result<PyDeserializedValue<'py>, DeserializationError> {
+        match v {
+            Some(v) => T::deserialize_py(typ, Some(v), py),
+            None => Ok(PyDeserializedValue::new(py.None().bind(py).to_owned())),
+        }
+    }
+}
+
+pub struct PyDeserializedValue<'py> {
+    pub value: Bound<'py, PyAny>,
+}
+
+impl<'py> PyDeserializedValue<'py> {
+    fn new(value: Bound<'py, PyAny>) -> Self {
+        Self { value }
+    }
+}
+impl<'py> IntoPyObject<'py> for PyDeserializedValue<'py> {
+    type Target = PyAny;
+    type Output = Bound<'py, PyAny>;
+    type Error = Infallible;
+    fn into_pyobject(self, _: Python<'py>) -> Result<Self::Output, Self::Error> {
+        Ok(self.value)
+    }
+}
+
+impl<'frame, 'metadata, 'py> PyDeserializeValue<'frame, 'metadata, 'py>
+    for PyDeserializedValue<'py>
+{
+    fn deserialize_py(
+        typ: &'metadata ColumnType<'metadata>,
+        v: Option<FrameSlice<'frame>>,
+        py: Python<'py>,
+    ) -> Result<Self, DeserializationError> {
+        let cql = deser_cql_py_value(py, typ, v)?;
+
+        Ok(PyDeserializedValue { value: cql })
+    }
+}
```

Comment:
:bulb: A-ha! So this is the main point of interest to us. All implementation on collections that I saw above are just to be used in `deser_cql_py_value`. Neat and reasonable :+1: